python/ml4ir/applications/ranking/features/feature_fns/normalization.py

```diff
@@ -0,0 +1,72 @@
+import tensorflow as tf
+
+from ml4ir.base.features.feature_fns.base import BaseFeatureLayerOp
+from ml4ir.base.io.file_io import FileIO
+
+
+class TheoreticalMinMaxNormalization(BaseFeatureLayerOp):
+    """
+    Min Max Normalization of individual query features,
+    where the theoretical min is used instead of the minimum.
+
+    Reference -> An Analysis of Fusion Functions for Hybrid Retrieval
+                 https://arxiv.org/abs/2210.11934
+    """
+    LAYER_NAME = "theoretical_min_max_norm"
+
+    THEORETICAL_MIN = "theoretical_min"
+
+
+    def __init__(self, feature_info: dict, file_io: FileIO, **kwargs):
+        """
+        Initialize layer to define a theoretical min max normalization
+
+        Parameters
+        ----------
+        feature_info : dict
+            Dictionary representing the configuration parameters for the specific feature from the FeatureConfig
+        file_io : FileIO object
+            FileIO handler object for reading and writing
+
+        Notes
+        -----
+        Args under feature_layer_info:
+            theoretical_min : float
+                Theoretical minimum to use for the query's record features
+                Default value of 0. is used if not specified.
+        """
+        super().__init__(feature_info=feature_info, file_io=file_io, **kwargs)
+
+        self.theoretical_min = self.feature_layer_args.get(self.THEORETICAL_MIN, 0.)
+
+
+    def call(self, inputs, training=None):
+        """
+        Defines the forward pass for the layer on the inputs tensor
+
+        Parameters
+        ----------
+        inputs: tensor
+            Input tensor on which the feature transforms are applied
+        training: boolean
+            Boolean flag indicating if the layer is being used in training mode or not
+
+        Returns
+        -------
+        tf.Tensor
+            Resulting tensor after the forward pass through the feature transform layer
+        """
+        # Replace values lower than theoretical minimum with theoretical minimum
+        inputs = tf.clip_by_value(inputs,
+                                  clip_value_min=self.theoretical_min,
+                                  clip_value_max=tf.reduce_max(inputs))
+
+        # Compute max values for each query
+        query_max = tf.expand_dims(tf.reduce_max(inputs, axis=-1), axis=-1)
+
+        # Min max normalization
```

Comment:
For min-max normalization you need to train the normalizer on your train dataset. Using a batch max value is not suitable for normalization. you need to write the max somewhere and read it when loading the normalizer. 
This is one reason to prefer Reciprocal Ranking that does not need any normalization.

Reply:
Not really, we are interested in only a per-query normalizer here.

Reply:
Indeed, in the paper they use per-query normalization. Something I don't understand is that this would give a score of 1 to the most relevant document of each result set. we are losing info of absolute similarities provided by the retrievers.  
it would be good to test whether using a p99 instead of min and max. It would mechanically provide the reranker robust information on which retriever models to favor rather than a rough estimate based on - for example - query length.

Reply:
> we are losing info of absolute similarities provided by the retrievers.
Yes, this is a trade off we have to do to deal with changing dense indices, unfortunately. Otherwise the model would expect a score of 0.8, say, and when it sees 0.9, it would be wrongly interpreting it.